flink-staging/flink-ml/src/main/scala/org/apache/flink/ml/optimization/GradientDescent.scala

```diff
@@ -0,0 +1,231 @@
+  * average gradient is then used to updated the weight values, including regularization.
+  *
+  * At the moment, the whole partition is used for SGD, making it effectively a batch gradient
+  * descent. Once a sampling operator has been introduced, the algorithm can be optimized
+  *
+  * @param runParameters The parameters to tune the algorithm. Currently these include:
+  *                      [[Solver.LossFunction]] for the loss function to be used,
+  *                      [[Solver.RegularizationType]] for the type of regularization,
+  *                      [[Solver.RegularizationParameter]] for the regularization parameter,
+  *                      [[IterativeSolver.Iterations]] for the maximum number of iteration,
+  *                      [[IterativeSolver.Stepsize]] for the learning rate used.
+  */
+class GradientDescent(runParameters: ParameterMap) extends IterativeSolver {
+
+  import Solver.WEIGHTVECTOR_BROADCAST
+
+  var parameterMap: ParameterMap = parameters ++ runParameters
+
+  /** Performs one iteration of Stochastic Gradient Descent using mini batches
+    *
+    * @param data A Dataset of LabeledVector (label, features) pairs
+    * @param currentWeights A Dataset with the current weights to be optimized as its only element
+    * @return A Dataset containing the weights after one stochastic gradient descent step
+    */
+  private def SGDStep(data: DataSet[(LabeledVector)], currentWeights: DataSet[WeightVector]):
+  DataSet[WeightVector] = {
+
+    // TODO: Sample from input to realize proper SGD
+    data.map {
+      new GradientCalculation
+    }.withBroadcastSet(currentWeights, WEIGHTVECTOR_BROADCAST).reduce {
+      (left, right) =>
+        val (leftGradVector, leftLoss, leftCount) = left
+        val (rightGradVector, rightLoss, rightCount) = right
+        // Add the left gradient to the right one
+        BLAS.axpy(1.0, leftGradVector.weights, rightGradVector.weights)
+        val gradients = WeightVector(
+          rightGradVector.weights, leftGradVector.intercept + rightGradVector.intercept)
+
+        (gradients , leftLoss + rightLoss, leftCount + rightCount)
+    }.map {
+      new WeightsUpdate
+    }.withBroadcastSet(currentWeights, WEIGHTVECTOR_BROADCAST)
+  }
+
+  /** Provides a solution for the given optimization problem
+    *
+    * @param data A Dataset of LabeledVector (label, features) pairs
+    * @param initWeights The initial weights that will be optimized
+    * @return The weights, optimized for the provided data.
+    */
+  override def optimize(
+    data: DataSet[LabeledVector],
+    initWeights: Option[DataSet[WeightVector]]): DataSet[WeightVector] = {
+    // TODO: Faster way to do this?
+    val dimensionsDS = data.map(_.vector.size).reduce((a, b) => b)
+
+    val numberOfIterations: Int = parameterMap(Iterations)
+
+    val initialWeightsDS: DataSet[WeightVector] = initWeights match {
+      case Some(x) => x
```

Comment:
Check that the provided weight vector is dense. If not, then make it dense. Alternatively, we can change the `WeightVector` type to only accept `DenseVector` weight vectors. What do you think?